subsys/bluetooth/host/hci_core.c

```diff
@@ -4786,16 +4789,18 @@ static void hci_vs_init(void)
 	int err;
 
 	/* If heuristics is enabled, try to guess HCI VS support by looking
-	 * at the HCI version and identity address. We haven't tried to set
-	 * a static random address yet at this point, so the identity will
-	 * either be zeroes or a valid public address.
+	 * at the HCI version and identity address. We haven't set any addresses
+	 * at this point. So we need to read the public address.
 	 */
 	if (IS_ENABLED(CONFIG_BT_HCI_VS_EXT_DETECT) &&
-	    (bt_dev.hci_version < BT_HCI_VERSION_5_0 ||
-	     (!atomic_test_bit(bt_dev.flags, BT_DEV_USER_ID_ADDR) &&
-	      bt_addr_le_cmp(&bt_dev.id_addr[0], BT_ADDR_LE_ANY)))) {
-		BT_WARN("Controller doesn't seem to support Zephyr vendor HCI");
-		return;
+	    (bt_dev.hci_version < BT_HCI_VERSION_5_0)) {
+		bt_addr_le_t addr;
+
+		if (!bt_read_public_addr(&addr)) {
```

Comment:
I'm not following the logic here. If the HCI version is less than 5.0 the controller will pretty certainly not support the Zephyr vendor extensions, so why bother sending the command? The vendor opcodes address space is freely definable per-vendor basis, so sending vendor commands can generate wildly unexpected results if the controller you're talking does support vendor commands but uses a completely different vendor command set. What makes things worse is that you're not checking the length of the returned parameters, so if this happens to map to a non-Zephyr command that returns a shorter set of parameters the code will read past the valid part of the buffer (in addition to this code falsely concluding that the Zephyr extensions are supported).

Reply:
You are right. I did this wrong.

Your point about length checking is very valid. It is not something I changed in this PR. But since I'm including new VS HCI commands I found it easiest to include a length check commit here.

I added length check when CONFIG_BT_HCI_VS_EXT_DETECT is used on all commands. Maybe it is a bit excessive but I found it safer than only checking on the first command complete for info. Might have been a coincidence.

Reply:
The heuristic should be that if the Version is 5.0 or later, and the controller does not have a valid public address. It's not a very good heuristic in my opinion. It works for this case because we only use to check for random static identity address. But in the case the controller supports many extensions to HCI but still has a valid Bluetooth address we would not be able to use these features.

Reply:
Yep, there really isn't any good solution here (if there was we wouldn't need a Kconfig to enable the heuristics). One thing we do in Linux is to check the reported manufacturer ID, however Zephyr allows overriding it so it's not particularly helpful.

Reply:
Hmm, maybe we should require the LF manufacturer ID by default for the heuristics? If someone wants to pair up Zephyr host with a controller with non-LF manufacturer ID then they won't be able to use the heuristics but just have the host assume that it's talking to a Zephyr?

Reply:
IIRC the Zephyr controller was introduced before LF had its SIG-assigned ID, which is why this code isn't taking advantage of it. Perhaps we can simply assume no-one is going to pair up a recent Zephyr host with an ancient Zephyr controller.

Reply:
I think the LF company ID is too restrictive. It would prevent us from having other controllers implement the Zephyr HCI specified VS extension. In my opinion the best approach would be to have a kind of magic number. But that would mean adding a read_magic command/event.

I will create an issue to handle this outside of this PR and leave it like it already was.